File: Regular/22-B.rb
#AtCoder Regular 22 B
#require 'pry'

length = STDIN.gets.to_i
taste  = STDIN.gets.chomp.split.map! {|s| s.to_i}

taste_longest = 0

h = Hash.new(0)

for i in 0...length
    #array = Array.new( 10 ** 5 ){ 0 }
    #h = Hash.new{ |hash,key| hash[key] = 0 }
    h.clear
    taste_long = 0
    #binding.pry
    for j in i...length
        if h[ taste[j] ] == 0 then
            h[ taste[j] ] = 1
            taste_long += 1
            #binding.pry
        else
            break
        end
    end
    if taste_longest <= taste_long then
        taste_longest = taste_long
    end
    if taste_longest >= ( length - i ) then
        break
    end
end
#binding.pry

puts taste_longest

File: Regular/22/22-B.c
//AtCoder Regular 22 B
#include <stdio.h>
#include <stdlib.h>

void clear_array( char *array , int length ){
    int i;
    for( i = 0; i < length; i++ )
        array[i] = 0;
}

int main( void ){
    int i;
    int j;
    int cursor = 0;
    int num;
    int taste_long = 0;
    int longest = 0;
    char buf[2048];
    char tmp[10];

    fscanf( stdin , "%d\n" , &num );
    //fscanf( stdin , "%s" , buf );
    fgets( buf , 2048 , stdin );
    //printf( "%s\n" , buf );

    int *taste = (int*)malloc( sizeof(int) * num );
    char *hash = (char*)malloc( sizeof(char) * num );

    for( i = 0; i < num * 2 - 1; ){
        clear_array( tmp , 10 );
        for( j = 0; j < 10; j++ ){
            if( buf[i] == '\0' || buf[i] == ' ' ){
                i++;
                break;
            }
            else{
                tmp[j] = buf[i];
                i++;
            }
        }
        taste[cursor] = atoi(tmp);
        cursor++;
    }

    //printf( "%d,%d\n",taste[0],taste[1] );

    for( i = 0; i < num; i++ ){
        clear_array( hash , num );
        taste_long = 0;
        for( j = i; j < num; j++ ){
            if( hash[ taste[j] ] == 0 ){
                hash[ taste[j] ] = 1;
                taste_long++;
            }
            else{
                break;
            }
        }
        if( longest <= taste_long ){
            longest = taste_long;
        }
    }

    printf( "%d\n" , longest );

    free( taste );
    free( hash );

    return 0;
}
